### ISSUE TITLE
Various tools fail under Linux 5.0

### ISSUE THREAD

**Issue by electrickite** [tpm2-software/tpm2-tools#1356]
After a kernel update to 5.0, various tpm2 tools have started failing. This is on an Arch Linux system, using the distro binary packages for `tpm2-{tss,tools}`. The same packages work as expected on kernel 4.20.13.

## Versions
```
tpm2-tools:  3.1.3
tpm2-tss:    2.2.0
uname:       Linux 5.0.0-arch1-1-ARCH #1 SMP PREEMPT Mon Mar 4 14:11:43 UTC 2019 x86_64 GNU/Linux
```

## Examples
```
$ sudo tpm2_nvlist 
0x1c00002:
WARNING:tcti:src/tss2-tcti/tcti-device.c:254:tcti_device_receive() Got EOF instead of response. 
ERROR: Reading the public part of the nv index failed with: 0xa0008
ERROR: Unable to run tpm2_nvlist

$ sudo tpm2_nvlist 
0x1c00002:
  hash algorithm:
    friendly: sha256
    value: 0xB
  attributes:
    friendly: ppwrite|writedefine|ppread|ownerread|authread|no_da|written|platformcreate
    value: 0x1200762
  size: 1177

0x1c0000a:
WARNING:tcti:src/tss2-tcti/tcti-device.c:254:tcti_device_receive() Got EOF instead of response. 
ERROR: Reading the public part of the nv index failed with: 0xa0008
ERROR: Unable to run tpm2_nvlist
```
```
$ sudo tpm2_nvread -x 0x01c00002
WARNING:tcti:src/tss2-tcti/tcti-device.c:254:tcti_device_receive() Got EOF instead of response. 
ERROR: Failed to query max transmission size viaTss2_Sys_GetCapability. Error:0xa0008
ERROR: Unable to run tpm2_nvread
```
```
$ sudo tpm2_getcap -c properties-fixed
WARNING:tcti:src/tss2-tcti/tcti-device.c:254:tcti_device_receive() Got EOF instead of response. 
ERROR: Failed to GetCapability: capability: 0x6, property: 0x100, TSS2_RC: 0xa0008

ERROR: Unable to run tpm2_getcap
```

**Comment by AndreasFuchsSIT** [comment]
Could you also post your kernel's TPM-related config attributes ?
I think to remember that linux 5.0 adds partial read/write on the devices. might be related to that.

**Comment by diabonas** [comment]
I can confirm these errors on my own Arch Linux system running kernel 5.0. The kernel configuration for the TPM of the [default Arch Linux kernel](https://git.archlinux.org/svntogit/packages.git/tree/trunk/config?h=packages/linux&id=9502fdcf546c933e5239231c8388131c7a346d50#n4199) is:
```
CONFIG_TCG_TPM=m
CONFIG_HW_RANDOM_TPM=y
CONFIG_TCG_TIS_CORE=m
CONFIG_TCG_TIS=m
CONFIG_TCG_TIS_SPI=m
CONFIG_TCG_TIS_I2C_ATMEL=m
CONFIG_TCG_TIS_I2C_INFINEON=m
CONFIG_TCG_TIS_I2C_NUVOTON=m
CONFIG_TCG_NSC=m
CONFIG_TCG_ATMEL=m
CONFIG_TCG_INFINEON=m
CONFIG_TCG_XEN=m
CONFIG_TCG_CRB=m
CONFIG_TCG_VTPM_PROXY=m
CONFIG_TCG_TIS_ST33ZP24=m
CONFIG_TCG_TIS_ST33ZP24_I2C=m
CONFIG_TCG_TIS_ST33ZP24_SPI=m
```
As you mentioned [partial read support in kernel version 5.0](https://github.com/torvalds/linux/commit/9488585b21bef0df1217e510c7134905d1d376a7), I tried building the current tpm2-tss master with any combination of the [`--enable-tcti-partial-reads`](https://github.com/tpm2-software/tpm2-tss/pull/1176) and [`--enable-tcti-device-async`](https://github.com/tpm2-software/tpm2-tss/pull/1121) configuration flags. The latter does not seem to change anything, the former still fails, but gives a different error message:
```
$ tpm2_nvlist
ERROR:tcti:src/tss2-tcti/tcti-device.c:176:tcti_device_receive() Failed to get response size fd 3, got errno 0: Success 
ERROR: GetCapability:Get NV Index list Error. TPM Error:0xa000a
ERROR: Unable to run /home/jonas/tpm2_nvlist
```
The same happens with the master version of tpm2-tools using the ESAPI:
```
tpm2_nvlist
0x1410001:
ERROR:tcti:src/tss2-tcti/tcti-device.c:176:tcti_device_receive() Failed to get response size fd 3, got errno 0: Success 
ERROR:esys:src/tss2-esys/api/Esys_NV_ReadPublic.c:310:Esys_NV_ReadPublic_Finish() Received a non-TPM Error 
ERROR:esys:src/tss2-esys/esys_tr.c:201:Esys_TR_FromTPMPublic_Finish() Error NV_ReadPublic ErrorCode (0x000a000a) 
ERROR:esys:src/tss2-esys/esys_tr.c:309:Esys_TR_FromTPMPublic() Error TR FromTPMPublic ErrorCode (0x000a000a) 
ERROR: Esys_TR_FromTPMPublic(0xA000A) - tcti:IO failure
ERROR: Failed to read the public part of NV index 0x1410001
ERROR: Unable to run tpm2_nvlist
```

**Comment by AndreasFuchsSIT** [comment]
can you also tell me, if you are running against /dev/tpmrm0 or /dev/tpm0 ?
Or if this make a difference or not ?

**Comment by diabonas** [comment]
Ah yes, I should have mentioned I use `/dev/tpmrm0`. Directly using `/dev/tpm0` does not seem to make a difference though, the errors stay the same both with and without `--enable-tcti-partial-reads` for tpm2-tools 3.1.3 and master.

Curiously, the command sometimes works partially or completely, so I get something like
```
$ tpm2_nvlist
0x1410001:
  hash algorithm:
    [...]

0x1410002:
ERROR:tcti:src/tss2-tcti/tcti-device.c:176:tcti_device_receive() Failed to get response size fd 3, got errno 0: Success
[...]
```
or even the complete list without any error if I am lucky. When this happens appears to be completely random, but it doesn't happen too often.

**Comment by AndreasFuchsSIT** [comment]
The lead towards the async being the question, aka if the whole answer is ready upon first read, then it's ok, if not, then not...

If you have the time, it would be interesting, if a sleep(5) at https://github.com/tpm2-software/tpm2-tss/blob/master/src/tss2-tcti/tcti-device.c#L119 would make a difference. That would more or less validate the async question. Otherwise, I'll do it sometime the coming week.

**Comment by diabonas** [comment]
With this additional delay between `tcti_device_context_cast` and `tcti_device_down_cast`, all the above commands as well as my own test cases work as expected without error (tested with `/dev/tpmrm0` and `tcti-device-async` and `tcti-partial-read` disabled as per default).

**Comment by electrickite** [comment]
Apologies for being absent on my own thread. Like @diabonas I am running the default Arch kernel, and the TPM kernel config is identical.

I also saw the same intermittent behavior: sometimes the commands will complete, although that is rare. More often, they will fail partway through their output and sometimes fail before producing any output at all.

**Comment by joshuagl** [comment]
This is being discussed on the linux-integrity list: https://www.spinics.net/lists/linux-integrity/msg06388.html

**Comment by AndreasFuchsSIT** [comment]
Maybe it was also a bug introduced with the partial reads stuff, since James' tools seem to work on the kernel list.
@diabonas: could you also try out branch 2.1.X of tpm2-tss ?
And thanks a lot for helping to triage the bug !

**Comment by diabonas** [comment]
No problem! The problem is present in 2.1.x as well, with and without `--enable-tcti-device-async`. Adding the delay helps on this branch as well. On the other hand, the 2.0.x branch is unaffected by the problem, everything works as expected there.

The issue is apparently related to https://github.com/tpm2-software/tpm2-tss/commit/3f2489b1c4572f203fb5c57b26364f9469ad716c: reverting that commit on 2.1.x causes all errors to disappear.

**Comment by AndreasFuchsSIT** [comment]
And there seems to be our first issue.
Even if `async` is disabled the open() uses O_NONBLOCK
https://github.com/tpm2-software/tpm2-tss/blob/master/src/tss2-tcti/tcti-device.c#L396

I would guess this O_NONBLOCK should need an `#ifdef` guard...

That could explain the error surrounding async disabled.
Why `--enable-tcti-device-async` errors is still a mystery however.

**Comment by diabonas** [comment]
Indeed, removing the `O_NONBLOCK` resolves the problem on the master and 2.1.x branches as long as `tcti-device-async` stays disabled.

Regarding the error when `tcti-device-async` is enabled, this might actually be a kernel bug as described in the [mailing list post](https://www.spinics.net/lists/linux-integrity/msg06388.html) that @joshuagl referred to. If I have the time, I might try testing the fix proposed there together with `--enable-tcti-device-async`.

However for the immediate problem, it should be enough to replace
```
    tcti_dev->fd = open (dev_path, O_RDWR | O_NONBLOCK);
```
by
```
#ifdef TCTI_ASYNC
    tcti_dev->fd = open (dev_path, O_RDWR | O_NONBLOCK);
#else
    tcti_dev->fd = open (dev_path, O_RDWR);
#endif
```
like you suggested. This fixes at least the default configuration of having `tcti-device-async` disabled with the current kernel 5.0.

**Comment by tstruk** [comment]
So it looks like this is a kernel regression. Lets send a fix for that.

**Comment by diabonas** [comment]
I confirm that the proposed fix for the kernel of replacing
```
	if (!priv->response_read || priv->response_length)
		mask = EPOLLIN | EPOLLRDNORM;
```
by
```
	if (priv->response_length)
		mask = EPOLLIN | EPOLLRDNORM;
```
in [`drivers/char/tpm/tpm-dev-common.c`](https://github.com/torvalds/linux/blob/master/drivers/char/tpm/tpm-dev-common.c) solves the problem. I tested the patch with a kernel based on [linux 5.0.1-arch1](https://git.archlinux.org/linux.git/log/?h=v5.0.1-arch1), which works with tpm2-tss 2.2.1 as well as the current master, both with `tcti-device-async` enabled and disabled.

**Comment by tstruk** [comment]
@diabonas right, the response_read prevents double writes (without) reading and should not be part of the poll condition. There is a time gap between a write which sets it to false and read (or timeout) that cleans it.

**Comment by redg3ar** [comment]
By now we are on 5.0.6, and I still see this error. I also see that it isnt changed in the linux-next tree as well. Any ideas when this will be patched?

**Comment by diabonas** [comment]
@redg3ar The [fix](https://git.kernel.org/pub/scm/linux/kernel/git/jmorris/linux-security.git/commit/?h=next-tpm&id=7bde1fe0abbe202dc1f484fe7ef110f6ed3b1351) has been [merged into the Linux kernel security system](https://www.spinics.net/lists/linux-integrity/msg06655.html), but it has not reached the main kernel tree yet.

**Comment by diabonas** [comment]
This is fixed in [Linux 5.0.10](https://lkml.org/lkml/2019/4/27/138).
